File: array/1337.go

type Item struct{
    socount int
    index int
}

type itemSlice []*Item

func kWeakestRows(mat [][]int, k int) []int {
    ret := make([]int, k)
    items := make(itemSlice, len(mat))
    for index, _ := range mat {
        count := 0
        for _, v := range mat[index] {
            if v == 1 {
                count++
            }else {
                break
            }
        }
        items[index] = &Item{count, index}
    }
    sort.Sort(items)
    for index, _ := range ret {
        ret[index] = items[index].index
    }
    return ret
}

func (t itemSlice) Len() int {
    return len(t)
}

func (t itemSlice) Less(i, j int) bool {
    if t[i].socount == t[j].socount {
        return t[i].index < t[j].index
    }else {
        return t[i].socount < t[j].socount
    }
}

func (t itemSlice) Swap(i, j int) {
    t[i], t[j] = t[j], t[i]
}

File: heap/1337.cc
class Solution {
struct Item {
    int count;
    int index;
};
public:
    vector<int> kWeakestRows(vector<vector<int>>& mat, int k) {
        std::vector<Item> v;
        for (int i = 0; i < mat.size(); ++i) {
            Item item;
            item.count = std::count(mat[i].begin(), mat[i].end(), 1);
            item.index = i;
            v.emplace_back(std::move(item));
        }
        std::sort(v.begin(), v.end(), [](Item& a, Item& b){
            if (a.count == b.count) return a.index < b.index;
            return a.count < b.count;
        });
        std::vector<int> ret;
        for (const auto& i : v) {
            if (k == 0) break;
            --k;
            ret.emplace_back(std::move(i.index));
        }
        return ret;
    }
};
